File: engines/python/agentce/domain.py
"""The domain ontology binding (SPEC §6.5): the enterprise's decision and context classes.

The binding subclasses ``agentce:Decision`` and ``agentce:ContextItem`` for the domain and declares
which decision types are consequential and the oversight modality each requires. The graph builder
uses it to materialise ``agentce:executesConsequential`` and ``agentce:oversightModalityMatchesDeclared``
and to extend the ``rdfs:subClassOf*`` closure. A run with no binding is valid (an empty binding).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

DECISION_ROOT = "agentce:Decision"
CONTEXT_ROOT = "agentce:ContextItem"


@dataclass
class DomainBinding:
    """A parsed domain binding."""

    #: child class -> parent class (domain subclasses of Decision / ContextItem).
    subclasses: dict[str, str] = field(default_factory=dict)
    #: decision-type IRIs declared consequential.
    consequential: set[str] = field(default_factory=set)
    #: decision-type IRI -> the oversight modality it requires.
    required_oversight: dict[str, str] = field(default_factory=dict)

    @classmethod
    def empty(cls) -> DomainBinding:
        return cls()

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> DomainBinding:
        binding = cls()
        for entry in data.get("decision_types", []) or []:
            if not isinstance(entry, dict) or "id" not in entry:
                continue
            decision_id = str(entry["id"])
            binding.subclasses[decision_id] = str(
                entry.get("subclass_of", DECISION_ROOT)
            )
            if entry.get("consequential"):
                binding.consequential.add(decision_id)
            modality = entry.get("required_oversight_modality")
            if isinstance(modality, str):
                binding.required_oversight[decision_id] = modality
        for entry in data.get("context_classes", []) or []:
            if isinstance(entry, dict) and "id" in entry:
                binding.subclasses[str(entry["id"])] = str(
                    entry.get("subclass_of", CONTEXT_ROOT)
                )
        return binding

    @classmethod
    def load(cls, path: Path) -> DomainBinding:
        data = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
        if not isinstance(data, dict):
            return cls.empty()
        return cls.from_dict(data)

File: engines/typescript/src/domain.ts
/**
 * The domain ontology binding (SPEC §6.5): the enterprise's decision and context classes.
 *
 * The binding subclasses `agentce:Decision` and `agentce:ContextItem` for the domain and declares
 * which decision types are consequential and the oversight modality each requires. A run with no
 * binding is valid (an empty binding).
 */

import { readFileSync } from "node:fs";
import { load } from "js-yaml";

export const DECISION_ROOT = "agentce:Decision";
export const CONTEXT_ROOT = "agentce:ContextItem";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export class DomainBinding {
  /** child class -> parent class (domain subclasses of Decision / ContextItem). */
  readonly subclasses = new Map<string, string>();
  /** decision-type IRIs declared consequential. */
  readonly consequential = new Set<string>();
  /** decision-type IRI -> the oversight modality it requires. */
  readonly requiredOversight = new Map<string, string>();

  static empty(): DomainBinding {
    return new DomainBinding();
  }

  static fromDict(data: Record<string, unknown>): DomainBinding {
    const binding = new DomainBinding();
    const decisionTypes = Array.isArray(data.decision_types) ? data.decision_types : [];
    for (const entry of decisionTypes) {
      if (!isRecord(entry) || !("id" in entry)) {
        continue;
      }
      const id = String(entry.id);
      binding.subclasses.set(
        id,
        String("subclass_of" in entry ? entry.subclass_of : DECISION_ROOT),
      );
      if (entry.consequential) {
        binding.consequential.add(id);
      }
      const modality = entry.required_oversight_modality;
      if (typeof modality === "string") {
        binding.requiredOversight.set(id, modality);
      }
    }
    const contextClasses = Array.isArray(data.context_classes) ? data.context_classes : [];
    for (const entry of contextClasses) {
      if (isRecord(entry) && "id" in entry) {
        binding.subclasses.set(
          String(entry.id),
          String("subclass_of" in entry ? entry.subclass_of : CONTEXT_ROOT),
        );
      }
    }
    return binding;
  }

  static load(path: string): DomainBinding {
    const data = load(readFileSync(path, "utf-8"));
    if (!isRecord(data)) {
      return DomainBinding.empty();
    }
    return DomainBinding.fromDict(data);
  }
}
